### PR TITLE
Program to remove empty cols and rows from CSV

### PR DESCRIPTION
**Pull request by SpencerHedger**
This program visits each file in an input directory and re-writes them to an output directory with empty columns and blank rows removed.

### CODE REVIEW

File: bin/remove_empty_rows_and_columns.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        basename = os.path.basename(filename)
+        out_filename = os.path.join(args.output_dir, basename)
+        with open(out_filename, 'w', encoding='utf-8-sig') as outfile:
+            with open(filename, newline='', encoding='utf-8-sig') as infile:
+                reader = csv.reader(infile)
+                headings = next(reader)
+                indicesToRemove = []
+                i = 0
+                for col in headings:
+                    if col == '':
```
Comment (mikejamesthompson) [inline bin/remove_empty_rows_and_columns.py:58]: You should probably trim whitespace before checking the heading is empty.

File: bin/remove_empty_rows_and_columns.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                        indicesToRemove.append(i)
+                    i = i + 1
+                writer = csv.writer(outfile)
+                removeIndices(headings, indicesToRemove)
+                writer.writerow(headings)
+
+                for line in reader:
+                    removeIndices(line, indicesToRemove)
```
Comment (mikejamesthompson) [inline bin/remove_empty_rows_and_columns.py:66]: Could you instead do:

```python
for index, line in enumerate(reader):
  if index in indicesToRemove:
    continue
    ...
```

And avoid the additional iteration?

Comment (mikejamesthompson) [reply bin/remove_empty_rows_and_columns.py:66]: Ignore this, I misunderstood :D

File: bin/remove_empty_rows_and_columns.py (hunk 1 above)
Comment (SpencerHedger) [reply bin/remove_empty_rows_and_columns.py]: Agreed, and doing this for data cells too.

File: bin/remove_empty_rows_and_columns.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                f'{directory} does not exist or is not a directory')
+
+    for filename in glob.glob(os.path.join(args.input_dir, '*.csv')):
+        basename = os.path.basename(filename)
+        out_filename = os.path.join(args.output_dir, basename)
+        with open(out_filename, 'w', encoding='utf-8-sig') as outfile:
+            with open(filename, newline='', encoding='utf-8-sig') as infile:
+                reader = csv.reader(infile)
+                headings = next(reader)
+                removeColumnsAfterIndex = len(headings)
```
Comment (duncanharris) [inline bin/remove_empty_rows_and_columns.py:51]: Should really be called `removeColumnsFromIndex`? i.e. my understanding is that columns at index and above are removed not the ones "after" index?

File: bin/remove_empty_rows_and_columns.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                                f' cell {len(headings)-i}')
+                            exit(1)
+                        removeColumnsAfterIndex = removeColumnsAfterIndex - 1
+                    else:
+                        seenNonEmptyColumn = True
+
+                writer = csv.writer(outfile)
+                writer.writerow(headings[:removeColumnsAfterIndex])
+
+                for lineNum, line in enumerate(reader):
```
Comment (duncanharris) [inline bin/remove_empty_rows_and_columns.py:68]: maybe `lineNum` should be named `dataLineIndex` and then assign `lineNum = dataLineIndex + 2`?

File: bin/remove_empty_rows_and_columns.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+                    # Keep this line.
+                    if contentLength > 0:
+                        writer.writerow(trimmedLine)
+                        if len(line) < removeColumnsAfterIndex:
+                            logging.warning(
+                                f'{basename}:{lineNum+2} too few cells on row: expected'
+                                f' {removeColumnsAfterIndex} but found {len(line)}')
+                    else:
+                        logging.info(
```
Comment (duncanharris) [inline bin/remove_empty_rows_and_columns.py:82]: Wondering if there going to be a lot of these and we should just count them up and log how many we removed?

**Review by duncanharris — CHANGES_REQUESTED** [review]

File: bin/remove_empty_rows_and_columns.py (hunk 5 above)
Comment (SpencerHedger) [reply bin/remove_empty_rows_and_columns.py:82]: Agreed, counting them up now and adding count to info log line for the file.

```
Read file from: Related_Datasets.csv and wrote modified file to: trim/Related_Datasets.csv: removed 3 columns and 6 rows
```

File: bin/remove_empty_rows_and_columns.py (hunk 4 above)
Comment (SpencerHedger) [reply bin/remove_empty_rows_and_columns.py:68]: Agreed.

File: bin/remove_empty_rows_and_columns.py (hunk 3 above)
Comment (SpencerHedger) [reply bin/remove_empty_rows_and_columns.py:51]: Agreed.

**Comment by SpencerHedger** [comment]
Had to make some further changes to variable names because snake case is required by linting checks.

File: REMOVE_EMPTY_ROWS_AND_COLUMNS.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+- Empty headings must be at the end, if the CSV file has an empty heading
+  amongst others this is considered to be an unrecoverable error
+
+Warnings will be logged (but processing will continue) when a data row:
+- contains fewer cells than expected
+- contains data in columns that are being removed
+
+## Example usage
+```
+python3 bin/remove_empty_rows_and_columns.py -i <PATH_TO_CSV_DIRECTORY> -o trimmed/
```
Comment (duncanharris) [inline REMOVE_EMPTY_ROWS_AND_COLUMNS.md:22]: does `trimmed` need a trailing `/`? Should we use `<PATH_TO_OUT_DIRECTORY>`

File: bin/remove_empty_rows_and_columns.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                        type=str,
+                        required=True,
+                        help='Input directory containing CSV files to transform')
+
+    parser.add_argument('-o', '--output-dir',
+                        type=str,
+                        required=True,
+                        help='Output directory to write modified files')
+
+    args = parser.parse_args()
```
Comment (duncanharris) [inline bin/remove_empty_rows_and_columns.py:36]: Wondering what happens if someone sets input and output directories to the same (or aliases of each other)?

**Review by duncanharris — APPROVED** [review]

File: bin/remove_empty_rows_and_columns.py (hunk 7 above)
Comment (duncanharris) [reply bin/remove_empty_rows_and_columns.py:36]: Thinking maybe we should insist that the output directory is empty. If we don't then they could end up with a mix of files we processed and files that are already there.

Comment (mikejamesthompson) [reply bin/remove_empty_rows_and_columns.py:36]: Should we enforce that it's totally empty? Or just empty of csv files? Might they conceivably want to have files of other types in there?

Comment (SpencerHedger) [reply bin/remove_empty_rows_and_columns.py:42]: Going with the `--force-overwrite` flag to allow the empty dir check to be skipped. Default is directory must be empty.

File: REMOVE_EMPTY_ROWS_AND_COLUMNS.md (hunk 6 above)
Comment (SpencerHedger) [reply REMOVE_EMPTY_ROWS_AND_COLUMNS.md]: Agreed - using `<PATH_TO_OUTPUT_DIRECTORY>`

**Pull request merged**